File: source/State/State.java
package source.State;

import com.badlogic.gdx.InputProcessor;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.math.Vector2;

/**
 * A state can be a Play source.State, Game Over source.State, Menu source.State, and so on.
 *
 * @author Xico
 */

public interface State extends InputProcessor {

    /**
     * Updates the state.
     * This is called 60 times per second.
     *
     * @param delta time elapsed since the last call to this method.
     */
    void update(float delta);

    /**
     * Renders the state.
     * This is also called 60 times per second.
     * Always called before {@link State#update(float)}.
     *
     * @param batch {@link Batch} used to render the state.
     */
    void render(Batch batch);

    /**
     * Pauses the state.
     */
    void pause();

    /**
     * Resumes the state, after it being paused.
     */
    void resume();

    /**
     * Hides the state.
     */
    void hide();

    /**
     * Disposes the state and all of the state's {@link com.badlogic.gdx.utils.Disposable} components.
     */
    void dispose();

    /**
     * Resizes the state and any of its components.
     *
     * @param width  new width.
     * @param height new height.
     */
    void resize(int width, int height);

    /**
     * Unprojects screen coordinates into world coordinates.
     *
     * @param screenX X in screen coordinates.
     * @param screenY Y in screen coordinates.
     * @return world coordinates.
     */
    Vector2 unproject(float screenX, float screenY);

    /**
     * Unprojects screen coordinates into world coordinates.
     *
     * @param screenCoordinates (X, Y) screen coordinates.
     * @return world coordinates.
     */
    Vector2 unproject(Vector2 screenCoordinates);

}

File: kot/state/State.kt
package kot.state

import com.badlogic.gdx.InputProcessor
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.math.Vector2

internal interface State : InputProcessor {

  /**
   * Updates the state.
   * This is called every frame.
   *
   * @param delta time elapsed since the last call to this method.
   */
  fun update(delta: Float)

  /**
   * Renders the state.
   * This is also called every frame, always after [State.update].
   *
   * @param batch [Batch] used to render the state.
   */
  fun render(batch: Batch)

  /**
   * Pauses the state.
   */
  fun pause()

  /**
   * Resumes the state, after it being paused.
   */
  fun resume()

  /**
   * Hides the state.
   */
  fun hide()

  /**
   * Disposes the state and all of the state's [com.badlogic.gdx.utils.Disposable] components.
   */
  fun dispose()

  /**
   * Resizes the state and any of its components.
   *
   * @param width  new width.
   * @param height new height.
   */
  fun resize(width: Int, height: Int)

  /**
   * Unprojects screen coordinates into world coordinates.
   *
   * @param screenX X in screen coordinates.
   * @param screenY Y in screen coordinates.
   * @return world coordinates.
   */
  fun unproject(screenX: Float, screenY: Float): Vector2

  /**
   * Unprojects screen coordinates into world coordinates.
   *
   * @param screenCoordinates (X, Y) screen coordinates.
   * @return world coordinates.
   */
  fun unproject(screenCoordinates: Vector2): Vector2

}
